include/pmacc/particles/algorithm/ForEach.hpp

```diff
@@ -0,0 +1,194 @@
+                mapper.getSuperCellIndex( DataSpace< dim >( blockIdx ) )
+            );
+
+            auto const & superCell = pb.getSuperCell( superCellIdx );
+            uint32_t const numPartcilesInSupercell = superCell.getNumParticles();
+
+
+            // end kernel if we have no particles
+            if( numPartcilesInSupercell == 0 )
+                return;
+
+            using FramePtr = typename T_ParBox::FramePtr;
+            FramePtr frame = pb.getLastFrame( superCellIdx );
+
+            // offset of the superCell (in cells, without any guards) to the origin of the local domain
+            DataSpace< dim > const localSuperCellOffset =
+                superCellIdx - mapper.getGuardingSuperCells( );
+
+            auto accFunctor = functor(
+                acc,
+                localSuperCellOffset,
+                WorkerCfg< T_numWorkers >{ workerIdx }
+            );
+
+            for( uint32_t parOffset = 0; parOffset < numPartcilesInSupercell; parOffset += frameSize)
+            {
+                using ParticleDomCfg = IdxConfig<
+                    frameSize,
+                    numWorkers
+                >;
+
+                // loop over all particles in the frame
+                ForEachIdx< ParticleDomCfg >{ workerIdx }(
+                    [&](
+                        uint32_t const linearIdx,
+                        uint32_t const
+                    )
+                    {
+                        // particle index within the supercell
+                        uint32_t parIdx = parOffset + linearIdx;
+                        auto particle = frame[ linearIdx ];
+
+                        bool const isPar = parIdx < numPartcilesInSupercell;
+                        if( isPar )
+                            accFunctor(
+                                acc,
+                                particle
+                            );
+                    }
+                );
+
+                frame = pb.getPreviousFrame( frame );
+            }
+        }
+    };
+
+} //namespace detail
+} //namespace acc
+
+    /** Run a unary functor for each particle of a species
+     *
+     * @warning Does NOT fill gaps automatically! If the
+     *          operation deactivates particles or creates "gaps" in any
+     *          other way, CallFillAllGaps needs to be called for the
+     *          species manually afterwards!
+     *
+     * Operates on the domain CORE and BORDER
+     *
+     * @tparam T_Functor unary particle functor type which follows the interface of
+     *                   pmacc::functor::Interface<F, 1u, void>
+     * @tparam T_Species type of the species
+     *
+     * @param species species to operate on
+     * @param functor operation which is applied to each particle of the species
+     */
+    template<
+        typename T_Functor,
+        typename T_Species
+    >
+    void forEach(
+        T_Species && species,
+        T_Functor && functor
```

Comment:
I think the forwarding reference for `functor` does not make sense since later we pass it by value to the kernel, and so an r-value with no copy constructor cannot be used here anyways. Would prefer passing it by value.

For `species` I think it's fine to keep `T_Species &&`.